Sources/ArgumentParser/Parsing/CommandParser.swift

```diff
@@ -145,14 +145,23 @@ extension CommandParser {
       .filter { prev in !decodedArguments.contains(where: { $0.type == prev.type })}
     decodedArguments.append(contentsOf: newDecodedValues)
     decodedArguments.append(DecodedArguments(type: currentNode.element, value: decodedResult))
+
+    return decodedResult
   }
   
   /// Starting with the current node, extracts commands out of `split` and
   /// descends into subcommands as far as possible.
   internal mutating func descendingParse(_ split: inout SplitArguments) throws {
     while true {
-      try parseCurrent(&split)
-      
+      var parsedCommand = try parseCurrent(&split)
+
+      // after decoding a command, make sure to validate it
+      do {
+        try parsedCommand.validate()
+      } catch {
+        throw CommandError(commandStack: [currentNode.element], parserError: ParserError.userValidationError(error))
```

Comment:
This needs to pass the full `commandStack` instead of just the current element, so that the error can have the right context when printing the usage information.